### PR TITLE
[SYCL] Move SYCL headers from standard clang location

### PR DESCRIPTION
**Pull request by vladimirlaz**
SYCL headers were moved from default clang include directory
(lib/clang/11.0.0/include) to (lib/clang/11.0.0/include/sycl) to
support mixed compilers (e.g. gcc for host code and clang for
device code) during build of SYCL application.

Clang driver adds new directory to system include search path.

For non-clang compilers the user should add option below to
compiler command line to let it find SYCL headers:
-I <path_to_sycl_compiler>/include/sycl

Additional diagnostic can be raised depending on compiler
which is used (see examples for clang below). It can be ignored.

/localdisk2/bbsycl/res/include/sycl/CL/__spirv/spirv_ops.hpp:74:8: error: SYCL 1.2.1 specification does not allow 'sycl_device' attribute applied to a function with a raw pointer parameter type [-Wsycl-strict]
  extern SYCL_EXTERNAL __ocl_event_t __spirv_GroupAsyncCopy(

/localdisk2/bbsycl/res/include/sycl/CL/sycl/ordered_queue.hpp:267:37: warning: 'ordered_queue' is deprecated: Replaced by in_order queue property [-Wdeprecated-declarations]
  size_t operator()(const cl::sycl::ordered_queue &q) const {

Signed-off-by: Vladimir Lazarev vladimir.lazarev@intel.com

### CODE REVIEW

**Comment by bader** [comment]
I suggest moving SYCL header out of compiler headers directory i.e. not put them to `lib/clang/11.0.0/include/`.
Can we move them to `include/sycl` directory?

File: sycl/test/basic_tests/buffer/buffer_interop.cpp [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out -L %opencl_libs_dir -lOpenCL
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out -I %sycl_include -L %opencl_libs_dir -lOpenCL
```
Comment (Fznamznon) [inline sycl/test/basic_tests/buffer/buffer_interop.cpp]: Why `-I %sycl_include` is added to the command line which contains `-fsycl` flag? Does it mean that user should add path to headers even if our compiler is invoked?

Comment (vladimirlaz) [reply sycl/test/basic_tests/buffer/buffer_interop.cpp]: it was done by a mistake.
removed

**Comment by vladimirlaz** [comment]
> could you please elaborate a bit more about this:
> 
> > Additional diagnostic can be raised depending on compiler
> > which is used (see examples for clang below). It can be ignored.
> > /localdisk2/bbsycl/res/lib/clang/11.0.0/include/sycl/CL/__spirv/spirv_ops.hpp:74:8: error: SYCL 1.2.1 specification does not allow 'sycl_device' attribute applied to a function with a raw pointer parameter type [-Wsycl-strict]
> > extern SYCL_EXTERNAL __ocl_event_t __spirv_GroupAsyncCopy(
> > /localdisk2/bbsycl/res/lib/clang/11.0.0/include/sycl/CL/sycl/ordered_queue.hpp:267:37: warning: 'ordered_queue' is deprecated: Replaced by in_order queue property [-Wdeprecated-declarations]
> > size_t operator()(const cl::sycl::ordered_queue &q) const {

As far as I understood compiler warnings are suppressed for system headers (found in path pointed by -internal-isystem).  Once the path is added with -I extra warnings were generated by the compiler. I mentined here examples from LIT execution.
It is up to user to suppress them.

**Comment by vladimirlaz** [comment]
> I suggest moving SYCL header out of compiler headers directory i.e. not put them to `lib/clang/11.0.0/include/`.
> Can we move them to `include/sycl` directory?

Both locations are possible but I prefer current location because:
 - I did not find good way to detect compiler root directory in clang driver. I will need to use .. symbols which might be error prone if there are symbolic links in the path. Current approach adds sub-folder which seems to be safer.
 - There was a request to preserve high level layout to decrease impact on tools and infrastructure.

**Comment by vzakhari** [comment]
> > > I suggest moving SYCL header out of compiler headers directory i.e. not put them to `lib/clang/11.0.0/include/`.
> > > Can we move them to `include/sycl` directory?
> > 
> > 
> > Both locations are possible but I prefer current location because:
> > 
> > * I did not find good way to detect compiler root directory in clang driver. I will need to use .. symbols which might be error prone if there are symbolic links in the path. Current approach adds sub-folder which seems to be safer.
> 
> Please take a look how standard library is handled.
> 
> > * There was a request to preserve high level layout to decrease impact on tools and infrastructure.
> 
> This is a terrible reason to implement error-prone design.

Do we want to keep versioning for sycl header files the same way clang does for its header files?  If yes, should the sycl header files' version match the clang version?

**Comment by bader** [comment]
> > > > I suggest moving SYCL header out of compiler headers directory i.e. not put them to `lib/clang/11.0.0/include/`.
> > > > Can we move them to `include/sycl` directory?
> > > 
> > > 
> > > Both locations are possible but I prefer current location because:
> > > 
> > > * I did not find good way to detect compiler root directory in clang driver. I will need to use .. symbols which might be error prone if there are symbolic links in the path. Current approach adds sub-folder which seems to be safer.
> > 
> > 
> > Please take a look how standard library is handled.
> > > * There was a request to preserve high level layout to decrease impact on tools and infrastructure.
> > 
> > 
> > This is a terrible reason to implement error-prone design.
> 
> Do we want to keep versioning for sycl header files the same way clang does for its header files? If yes, should the sycl header files' version match the clang version?

Maybe, but I don't think that moving library headers to the compiler sub-directory is the right approach or has any connection to the library versioning.
SYCL runtime library is a separate project and I prefer treating it that way.
E.g. libc++ is another LLVM project, which versioning following LLVM scheme, but libc++ headers are not installed into compiler built-in headers sub-directory. libc++ project uses CMake utilities to align versioning with the rest of the project and we might do the same.

**Comment by bader** [comment]
@vladimirlaz, please, resolve merge conflict and fix tests.

**Comment by vladimirlaz** [comment]
> @vladimirlaz, please, resolve merge conflict and fix tests.

done

**Comment by bader** [comment]
@vladimirlaz, OpenCL headers not found during CTS compilation.

File: sycl/CMakeLists.txt [hunk 2]
```diff
@@ -54,9 +54,9 @@ endif()
 # Unlike PACKAGE_VERSION, CLANG_VERSION does not include LLVM_VERSION_SUFFIX.
 set(CLANG_VERSION "${CLANG_VERSION_MAJOR}.${CLANG_VERSION_MINOR}.${CLANG_VERSION_PATCHLEVEL}")
 
-set(LLVM_INST_INC_DIRECTORY "lib${LLVM_LIBDIR_SUFFIX}/clang/${CLANG_VERSION}/include")
-set(dst_dir ${LLVM_LIBRARY_OUTPUT_INTDIR}/clang/${CLANG_VERSION}/include)
-set(dst_deploy_dir ${CMAKE_INSTALL_PREFIX}/lib/clang/${CLANG_VERSION}/include)
+set(SYCL_INST_INC_DIRECTORY "include/sycl")
```
Comment (bader) [inline sycl/CMakeLists.txt]:
```suggestion
set(SYCL_INCLUDE_DIR "include/sycl")
```

File: sycl/CMakeLists.txt [hunk 3]
```diff
@@ -54,9 +54,9 @@ endif()
...
+set(dst_dir ${LLVM_BINARY_DIR}/${SYCL_INST_INC_DIRECTORY})
```
Comment (bader) [inline sycl/CMakeLists.txt]:
```suggestion
set(CL_HEADERS_DIR ${LLVM_BINARY_DIR}/${SYCL_INST_INC_DIRECTORY})
```

**Pull request merged**